### DISCUSSION TITLE
GUI freezes when trying to Select and Overwrite an existing file with the RFD crate. (Windows)

### DISCUSSION THREAD

**Discussion by Munjen** [slint-ui/slint#4102] [Q&A]
Hello!
I'm using the rfd crate to create a Save Dialog for my slint application on Windows 10.
The Save Dialog works perfectly except when I select an **existing file** and click the **save** button (on the save dialog).
**The Save Dialog stays open and both the Slint UI and the Save Dialog lock up refusing to respond to buttons and controls.**
Nothing can stop the process except 'kill terminal' in VSCode.

A simple test project **with just the rfd crate** ...and Save Dialog works fine.
The overwrite file message box opens and the Save Dialog closes if user responds to 'Overwrite existing file?' with 'Yes' as expected!

I put the rfd code inside a closure that runs when a slint save button is pressed
I'm not sure if this is some bug in slint, rfd or am I doing something incorrectly?

EDIT: Today I tried compiling on Linux Mint 21.2, Cinnamon v5.8.4, Linux Kernel: 5.15.0-88-generic.
**On Linux everything works fine, No issues at all!** I will post this issue to RFD git as well because now it seems more likely that the problem is coming from the rfd crate.

```
main_window.on_save_pressed(move |x, y| {
        let file = match save_dialog() {
            Some(s) => s,
            None => return,
        };

       //... doing stuff with the returned string...
       //... doing stuff with x, y parameters...
       //... doing stuff with a weak pointer to main_window...
});

use rfd::FileDialog;
fn save_dialog() -> Option<String> {
    let file = FileDialog::new()
        .set_file_name("Output.exr")
        .add_filter("Open EXR", &["exr"])
        .set_title("Save As 32bit EXR")
        .save_file();

    match file {
        Some(f) => Some(f.into_os_string().into_string().unwrap()),
        None => None,
    }
}
```

**Comment by ogoffart** [comment]
(I converted the issue into a discussion to keep our bug tracker for actual bugs and feature request)

In order to debug that, you can try to attach a debugger to the process and try to find out what its doing by getting a backtrace for all threads.

Is one thread using 100% of the CPU?

**Reply by Munjen** [reply]
Thank you for the reply!

Sorry I did not have the time to reply sooner. One of the CPU cores remains at about 70-80% once the rfd dialog locks.
The Slint window can still be moved around and put on front by clicking on it but Slint controls don't respond to input.
I don't know how to attach a Debugger cause I'm relatively new to Rust programming and low level programming in general. Mostly worked with Python until recently. I will post more information once I get the time to learn more.
It looks like rfd itself should be the cause of the problem. I posted a bug report there as well but still no reply. Also I tried clearing everything with cargo clean and recompiling but it did not help after all. (once that worked for a particular case -> probably some files got corrupted)

**Reply by Munjen** [reply]
With my limited skills I tried to run in debug mode with VSCode and the farthest I can trace it are these two functions:
rfd-0.12.1\src\backend\win_cid\utils.rs --> init_com function on line 28. calls
rfd-0.12.1\src\backend\win_cid\file_dialog.rs --> line 106: impl FileSaveDialogImpl for FileDialog --> line 111: dialog.show()?;

The dialog opens on line 111 and becomes interactive. But after pressing save to overwrite an existing file everything freezes. Probably it is some unsafe interface with windows API or something causing the problem.

**Reply by ogoffart** [reply]
maybe then it is a bug in RFD.
Can you reproduce the problem without Slint?

**Reply by asuper0** [reply]
The rfd dialog is just not shown, or is behind the slint window. It can be displayed to the front by press 'Alt'.
This is strange, they works fine with slint 1.1.1 and rfd 0.11.4, in one of my program. I tried slint 1.3.2 and rfd 0.11.4 which could not work as well.
My current code like this:
```rust
window.on_apply_config(
|| {
  // some code
  rfd::MessageDialog::new()
      .set_title("Warning")
      .set_description(&msg.to_string())
      .show();
}
```

**Reply by asuper0** [reply]
I test again today, this happens with slint >=1.3.0, while slint <=1.2.2 doesn't.
This could be reproduced with very simple code
```rust
// [dependencies]
// rfd = "0.12.1"
use rfd::MessageDialog;
slint::slint!{
    import { Button } from "std-widgets.slint";
    export component MainWindow inherits Window{
        width: 100px;
        height: 80px;
        callback show-dlg();
        Button { text: "button"; clicked => { show-dlg(); }}
    }
}

fn main() {
    let window = MainWindow::new().unwrap();

    window.on_show_dlg(|| {
        MessageDialog::new().set_description("hello").show();
    });
    window.run().unwrap();
}

```

When click the button, the windows system will play the sound which means a message box is shown, but is not really shown.
This is the same as the "save file dialog" when the selected file is existed.
Both of dialogs above could be shown by pressing "Alt".

**Reply by oswald2** [reply]
I have the same issue. Seems like the dialog is hidden somewhere. Alt-Tab away and back to the slint application also works, that is how I currently informed users to mitigate.

**Reply by asuper0** [reply]
Please take a look if you have time @ogoffart 😀
As metioned above, this happens with slint >=1.3.0, while slint <=1.2.2 doesn't.

**Reply by ogoffart** [reply]
I haven't had time to have a look at that yer.
One main difference between Slint 1.3 and 1.2 is the version of winit which was bumped to 0.29

**Comment by RutekFurier** [comment]
In my case, I used https://github.com/slint-ui/slint-rust-template template project and tried to create a file picker dialog which resulted in the same issue (app freezing). I tried to force downgrade slint to "1.2" and "1.1" but it didn't fix the issue. Only downgrading rfd from "0.13.0" to "0.12.0" worked

**Reply by RutekFurier** [reply]
I should probably mention that I'm using Linux since the issue mentions Windows

**Comment by heipiao233** [comment]
Try to run it on a separated thread?
